programs/stake/src/stake_state.rs

```diff
@@ -132,6 +132,85 @@ fn redelegate_stake(
     Ok(())
 }
 
+fn move_stake_or_lamports_shared_checks(
+    invoke_context: &InvokeContext,
+    transaction_context: &TransactionContext,
+    instruction_context: &InstructionContext,
+    source_account: &BorrowedAccount,
+    lamports: u64,
+    destination_account: &BorrowedAccount,
+    clock: &Clock,
+    stake_history: &StakeHistory,
+    stake_authority_index: IndexOfAccount,
+) -> Result<(MergeKind, MergeKind), InstructionError> {
+    // authority must sign
+    let stake_authority_pubkey = transaction_context.get_key_of_account_at_index(
+        instruction_context
+            .get_index_of_instruction_account_in_transaction(stake_authority_index)?,
+    )?;
+    if !instruction_context.is_instruction_account_signer(stake_authority_index)? {
+        return Err(InstructionError::MissingRequiredSignature);
+    }
+
+    let mut signers = HashSet::new();
+    signers.insert(*stake_authority_pubkey);
+
+    // check owners
+    if *source_account.get_owner() != id() || *destination_account.get_owner() != id() {
+        return Err(InstructionError::IncorrectProgramId);
+    }
+
+    // confirm not the same account
+    if *source_account.get_key() == *destination_account.get_key() {
+        return Err(InstructionError::InvalidInstructionData);
+    }
+
+    // source and destination must be writable
+    if !source_account.is_writable() || !destination_account.is_writable() {
+        return Err(InstructionError::InvalidInstructionData);
+    }
```

Comment:
I think we can just depend on the runtime to enforce this, can we not?

Reply:
i waffled on whether this is pointless or not but decided to include it because theres a bullet point in the simd saying non-writable accounts == transaction aborts, which implies this happens even if no attempt to write to them is made, and while i dont think there are any code paths that could result in an `Ok(())` no-op without a write attempt, it felt safer

Reply:
Fair enough. There's no real harm to the redundancy; just extra compute when we get to the BPF version, I guess, which we can reassess then.